vllm/model_executor/model_loader.py

```diff
@@ -77,7 +80,11 @@ def get_model(model_config: ModelConfig, device_config: DeviceConfig,
         # Create a model instance.
         # The weights will be initialized as empty tensors.
         with torch.device(device_config.device):
-            if hasattr(model_class, "supported_lora_modules"):
+            if model_config.load_format == "tensorizer" and _is_vllm_model(
+                    model_config):
+                model = load_with_tensorizer(model_class, model_config)
+                return model.eval()
+            elif hasattr(model_class, "supported_lora_modules"):
                 model = model_class(model_config.hf_config, linear_method,
                                     lora_config)
```

Comment:
Using Tensorizer and LoRA should be able to both be used.

The code block you're highlighting is a specific instance of using Tensorizer using `load_with_tensorizer`, which skips loading a model through its instantiation as a vLLM model (not loading a model from HuggingFace and loading it with vLLM's bespoke layer modules). It assumes that the model to be deserialized _is itself already_ a vLLM model. My example script `examples/tensorize_vllm_model.py` shows an example of serializing a model at a state that is ready to do inference on vLLM. I'm not sure how that interacts with LoRA adapters in that case.

Tensorizer can additionally be invoked the "standard" way, like any other load format, through its implementation in `weight_utils.py`, and using serialized models from the HF model hub. In this instance, LoRA adapters should still work, as it's simply loading in the tensors and handing it off to vLLM's `weight_loader`s as normal. Looking at your example script for multiLoRA inference, I was able to successfully run it using Tensorizer with this modification:

```
def initialize_engine() -> LLMEngine:
    """Initialize the LLMEngine."""
    # max_loras: controls the number of LoRAs that can be used in the same
    #   batch. Larger numbers will cause higher memory usage, as each LoRA
    #   slot requires its own preallocated tensor.
    # max_lora_rank: controls the maximum supported rank of all LoRAs. Larger
    #   numbers will cause higher memory usage. If you know that all LoRAs will
    #   use the same rank, it is recommended to set this as low as possible.
    # max_cpu_loras: controls the size of the CPU LoRA cache.
    from vllm.model_executor.tensorizer_loader import TensorizerArgs
    engine_args = EngineArgs(model="meta-llama/Llama-2-7b-hf",
        load_format="tensorizer",
        enable_lora=True,
        tensorizer_args=TensorizerArgs(
            tensorizer_uri=tensorizer_uri,
            num_readers=1),
        max_loras=1,
        max_lora_rank=8,
        max_cpu_loras=2,
        max_num_seqs=50 ## Lowered for GPU usage,
        max_model_len=1000 ## Lowered for GPU usage)
```

Reply:
Thanks! Just to be clear, does it mean that in an ideal case, we'll load directly to GPU?

For LoRA, the underlying layers have exactly the same weights, the only difference is that the LoRA layers instantiate some extra tensors (there is no extra data being loaded).

Reply:
> Just to be clear, does it mean that in an ideal case, we'll load directly to GPU?

Yes, ideally Tensorizer can be used to have CUDA be the output device. The loading mechanism for vLLM seems to be initializing the model entirely in CPU memory, and then copying it into GPU memory. `tensorizer` achieves this much faster and with far less CPU memory usage by instantiating a few reusable CPU buffers, and the model with meta tensors when copying to CUDA devices rather than instantiating everything first into memory.